File: solutions/python/successor.py
from binary_tree_with_parent_prototype import BinaryTreeNode


# @include
def find_successor(node):
    if node.right:
        # Successor is the leftmost element in node's right subtree.
        node = node.right
        while node.left:
            node = node.left
        return node

    # Find the closest ancestor whose left subtree contains node.
    while node.parent and node.parent.right is node:
        node = node.parent

    # A return value of None means node does not have successor, i.e., node is
    # the rightmost node in the tree.
    return node.parent
# @exclude


def main():
    #      3
    #    2   5
    #  1    4 6
    root = BinaryTreeNode(3)
    root.parent = None
    assert not find_successor(root)
    root.left = BinaryTreeNode(2)
    root.left.parent = root
    assert find_successor(root.left).data == 3

    root.left.left = BinaryTreeNode(1)
    root.left.left.parent = root.left
    assert find_successor(root.left).data == 3
    assert find_successor(root.left.left).data == 2

    root.right = BinaryTreeNode(5)
    root.right.parent = root
    root.right.left = BinaryTreeNode(4)
    root.right.left.parent = root.right
    root.right.right = BinaryTreeNode(6)
    root.right.right.parent = root.right
    # should output 6
    node = find_successor(root.right)
    assert 6 == node.data
    if node:
        print(node.data)
    else:
        print('null')
    # should output 'null'
    node = find_successor(root.right.right)
    assert not node
    if node:
        print(node.data)
    else:
        print('null')


if __name__ == '__main__':
    main()

File: solutions/cpp/successor.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Binary_tree_with_parent_prototype.h"

using std::cout;
using std::endl;
using std::make_unique;
using std::unique_ptr;

// @include
BinaryTreeNode<int>* FindSuccessor(
    const unique_ptr<BinaryTreeNode<int>>& node) {
  auto* iter = node.get();
  if (iter->right != nullptr) {
    // Successor is the leftmost element in node's right subtree.
    iter = iter->right.get();
    while (iter->left) {
      iter = iter->left.get();
    }
    return iter;
  }

  // Find the closest ancestor whose left subtree contains node.
  while (iter->parent != nullptr && iter->parent->right.get() == iter) {
    iter = iter->parent;
  }
  // A return value of nullptr means node does not have successor, i.e., it is
  // the rightmost node in the tree.
  return iter->parent;
}
// @exclude

int main(int argc, char* argv[]) {
  //      3
  //    2   5
  //  1    4  6
  auto root = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{3});
  root->parent = nullptr;
  assert(FindSuccessor(root) == nullptr);
  root->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{2});
  root->left->parent = root.get();
  assert(FindSuccessor(root->left)->data == 3);

  root->left->left = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{1});
  root->left->left->parent = root->left.get();
  assert(FindSuccessor(root->left)->data == 3);
  assert(FindSuccessor(root->left->left)->data == 2);

  root->right = make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{5});
  root->right->parent = root.get();
  root->right->left =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{4});
  root->right->left->parent = root->right.get();
  root->right->right =
      make_unique<BinaryTreeNode<int>>(BinaryTreeNode<int>{6});
  root->right->right->parent = root->right.get();
  // should output 6
  auto* node = FindSuccessor(root->right);
  assert(6 == node->data);
  if (node) {
    cout << node->data << endl;
  } else {
    cout << "null" << endl;
  }
  // should output "null"
  node = FindSuccessor(root->right->right);
  assert(!node);
  if (node) {
    cout << node->data << endl;
  } else {
    cout << "null" << endl;
  }
  return 0;
}
